File: api/schemas/outcomes.py
"""Stable Outcomes / QA schemas for internal voice-ops tools (P0)."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class QaNodeOutcome(BaseModel):
    """Normalized single-node (or whole-call) QA result."""

    node_id: str
    node_name: str = ""
    score: float | None = None
    tags: list[str] = Field(default_factory=list)
    summary: str = ""
    sentiment: str | None = None  # positive|neutral|negative|…
    error: str | None = None
    raw: dict[str, Any] = Field(default_factory=dict)


class QaRunOutcome(BaseModel):
    """Normalized QA payload for one workflow run."""

    schema_version: Literal[1] = 1
    run_id: int
    workflow_id: int | None = None
    has_qa: bool = False
    overall_score: float | None = None
    sentiment: str | None = None
    tags: list[str] = Field(default_factory=list)
    nodes: list[QaNodeOutcome] = Field(default_factory=list)
    errors: list[str] = Field(default_factory=list)
    source_keys: list[str] = Field(default_factory=list)


class OutcomeRunRow(BaseModel):
    run_id: int
    workflow_id: int
    workflow_name: str = ""
    created_at: datetime | None = None
    is_completed: bool = False
    disposition: str = "UNKNOWN"
    phone_number: str = ""
    duration_seconds: float | None = None
    call_tags: list[str] = Field(default_factory=list)
    qa: QaRunOutcome
    campaign_id: int | None = None


class AggregationSampleMeta(BaseModel):
    """Shared sample/truncation fields for in-memory aggregates."""

    total_matching_runs: int = 0
    sampled_runs: int = 0
    sample_limit: int = 0
    truncated: bool = False
    truncation_note: str | None = None


class OutcomesSummaryResponse(BaseModel):
    from_date: str
    to_date: str
    timezone: str
    workflow_id: int | None = None
    campaign_id: int | None = None
    total_runs: int
    completed_runs: int
    disposition_distribution: list[dict[str, Any]]
    qa_coverage: dict[str, Any]
    average_qa_score: float | None = None
    top_qa_tags: list[dict[str, Any]]
    total_matching_runs: int = 0
    sampled_runs: int = 0
    sample_limit: int = 0
    truncated: bool = False
    truncation_note: str | None = None


class OutcomesListResponse(BaseModel):
    total: int
    page: int
    limit: int
    runs: list[OutcomeRunRow]

File: ui/src/lib/api/outcomes.ts
/**
 * Typed Outcomes API client (P0).
 *
 * OpenAPI codegen requires a live backend (`npm run generate-client`).
 * Until the full client is regenerated against a running API with the new
 * routes, this module is the stable typed surface for `/api/v1/outcomes/*`.
 */

import { client } from "@/client/client.gen";

export type QaNodeOutcome = {
  node_id: string;
  node_name: string;
  score: number | null;
  tags: string[];
  summary: string;
  error?: string | null;
  raw?: Record<string, unknown>;
};

export type QaRunOutcome = {
  schema_version: 1;
  run_id: number;
  workflow_id: number | null;
  has_qa: boolean;
  overall_score: number | null;
  tags: string[];
  nodes: QaNodeOutcome[];
  errors: string[];
  source_keys: string[];
};

export type OutcomeRunRow = {
  run_id: number;
  workflow_id: number;
  workflow_name: string;
  created_at: string | null;
  is_completed: boolean;
  disposition: string;
  phone_number: string;
  duration_seconds: number | null;
  call_tags: string[];
  qa: QaRunOutcome;
};

export type DispositionBucket = {
  disposition: string;
  count: number;
  percentage: number;
};

export type OutcomesSummary = {
  from_date: string;
  to_date: string;
  timezone: string;
  workflow_id: number | null;
  total_runs: number;
  completed_runs: number;
  disposition_distribution: DispositionBucket[];
  qa_coverage: {
    runs_with_qa: number;
    runs_without_qa: number;
    coverage_pct: number;
  };
  average_qa_score: number | null;
  top_qa_tags: Array<{ tag: string; count: number }>;
};

export type OutcomesListResponse = {
  total: number;
  page: number;
  limit: number;
  runs: OutcomeRunRow[];
};

export type OutcomesQuery = {
  from_date: string;
  to_date: string;
  timezone?: string;
  workflow_id?: number | null;
  page?: number;
  limit?: number;
};

function errorMessage(err: unknown, fallback: string): string {
  if (typeof err === "object" && err && "detail" in err) {
    return String((err as { detail: unknown }).detail);
  }
  if (err instanceof Error) return err.message;
  return fallback;
}

export async function fetchOutcomesSummary(
  q: OutcomesQuery,
): Promise<OutcomesSummary> {
  const res = await client.get({
    url: "/api/v1/outcomes/summary",
    query: {
      from_date: q.from_date,
      to_date: q.to_date,
      timezone: q.timezone || "UTC",
      ...(q.workflow_id != null ? { workflow_id: q.workflow_id } : {}),
    },
  });
  if (res.error) {
    throw new Error(errorMessage(res.error, "Outcomes summary failed"));
  }
  return res.data as OutcomesSummary;
}

export async function fetchOutcomesRuns(
  q: OutcomesQuery,
): Promise<OutcomesListResponse> {
  const res = await client.get({
    url: "/api/v1/outcomes/runs",
    query: {
      from_date: q.from_date,
      to_date: q.to_date,
      timezone: q.timezone || "UTC",
      page: q.page ?? 1,
      limit: q.limit ?? 50,
      ...(q.workflow_id != null ? { workflow_id: q.workflow_id } : {}),
    },
  });
  if (res.error) {
    throw new Error(errorMessage(res.error, "Outcomes runs failed"));
  }
  return res.data as OutcomesListResponse;
}

export async function fetchRunQa(runId: number): Promise<QaRunOutcome> {
  const res = await client.get({
    url: `/api/v1/outcomes/runs/${runId}/qa`,
  });
  if (res.error) {
    throw new Error(errorMessage(res.error, "QA fetch failed"));
  }
  return res.data as QaRunOutcome;
}
